icu4j/main/classes/core/src/com/ibm/icu/message2/FormattedPlaceholder.java

```diff
@@ -0,0 +1,76 @@
+// © 2022 and later: Unicode, Inc. and others.
+// License & terms of use: http://www.unicode.org/copyright.html
+
+package com.ibm.icu.message2;
+
+import com.ibm.icu.text.FormattedValue;
+
+/**
+ * A richer formatting result, encapsulating a {@link FormattedValue}, the original value to format,
+ * and we are considering adding some more info.
+ * Very preliminary.
+ *
+ * @internal ICU 72 technology preview
+ * @deprecated This API is for ICU internal use only.
+ */
+@Deprecated
+public class FormattedPlaceholder {
+    private final FormattedValue formattedValue;
+    private final Object inputValue;
+
+    /**
+     * Constructor creating the immutable {@code FormattedPlaceholder}.
+     *
+     * @param formattedValue the result of formatting the placeholder.
+     * @param inputValue the original value to be formatted.
+     *
+     * @internal ICU 72 technology preview
+     * @deprecated This API is for ICU internal use only.
+     */
+    @Deprecated
+    public FormattedPlaceholder(FormattedValue formattedValue, Object inputValue) {
+        this.formattedValue = formattedValue;
+        this.inputValue = inputValue;
+    }
+
+    /**
+     * Retrieve the original input value that was formatted.
+     *
+     * @return the original value to be formatted.
+     *
+     * @internal ICU 72 technology preview
+     * @deprecated This API is for ICU internal use only.
+     */
+    @Deprecated
+    public Object getInput() {
+        return inputValue;
+    }
+
+    /**
+     * Retrieve the formatted value.
+     *
+     * @return the result of formatting the placeholder.
+     *
+     * @internal ICU 72 technology preview
+     * @deprecated This API is for ICU internal use only.
+     */
+    @Deprecated
+    public FormattedValue getFormattedValue() {
+        return formattedValue;
+    }
+
+    /**
+     * Returns a string representation of the object.
+     * It can be null, which is unusual, and we plan to change that.
+     *
+     * @return a string representation of the object.
+     *
+     * @internal ICU 72 technology preview
+     * @deprecated This API is for ICU internal use only.
+     */
+    @Deprecated
+    @Override
+    public String toString() {
+        return formattedValue == null ? null : formattedValue.toString();
```

Comment:
Kind of yes.

I tried to emulate what MessageFormat does.
When something goes wrong (for certain kinds of "wrong" :-) MF1 returns the placeholder un-replaced.

For example `"Hello {user}"`, if there was no parameter named "user", the result is `"Hello {user}"`
So it does not throw.

Since the functions in placeholders don't have access to their own name, or to the name of the argument they format (only the value of the arg to format), there is no way for a function that goes wrong to
"fallback" to "...{$foo}..." or "...{:bar}...".

But the the "engine" knows that. The function returns null in the formatted value, and the engine
decides what to do (throw, show `null` or `{$foo}`).

Plus we get consistency: if all functions return null when they fail, all failing placeholders look the same.
You don't get empty string from one, "null" from another, and so on ("Hello null, tomorrow is {$date}")

Reply:
TODO: FormattedValue should never be null.
In case of error the functions should throw.

Reply:
Try doing it.

Reply:
Done.
A lot less painful than I was afraid of.
Only one test failed, the JS one with `"hello {null}"`

Throwing a NPE as this is what MF1 does, but let me know if you would rather see `IllegalArgumentException`.

Reply:
Now that we no longer allow a null formattedValue, please stop checking for it.
```suggestion
        return formattedValue.toString();
```

Reply:
Done.